File: AugmentedRealityWithArucoMarkers/generate_aruco_markers.py
#!/usr/bin/env python3
"""Generate a DICT_6X6_250 ArUco marker with current OpenCV."""

from __future__ import annotations

import argparse
from pathlib import Path
import sys

import cv2 as cv
import numpy as np


PROJECT_DIR = Path(__file__).resolve().parent


def generate_marker(
    marker_id: int = 33,
    size: int = 200,
    border_bits: int = 1,
) -> np.ndarray:
    if not 0 <= marker_id < 250:
        raise ValueError("marker_id must be between 0 and 249.")
    if size < 32:
        raise ValueError("size must be at least 32 pixels.")
    if border_bits < 1:
        raise ValueError("border_bits must be at least 1.")

    dictionary = cv.aruco.getPredefinedDictionary(cv.aruco.DICT_6X6_250)
    return dictionary.generateImageMarker(
        marker_id, size, borderBits=border_bits
    )


def parse_args(argv: list[str] | None = None) -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Generate a 6x6 ArUco marker.")
    parser.add_argument("--id", type=int, default=33, help="Marker ID from 0 to 249.")
    parser.add_argument(
        "--size", type=int, default=200, help="Square image size in pixels."
    )
    parser.add_argument(
        "--border-bits",
        type=int,
        default=1,
        help="Marker border width in bits.",
    )
    parser.add_argument(
        "--output",
        type=Path,
        default=PROJECT_DIR / "marker33.png",
        help="Output PNG path.",
    )
    return parser.parse_args(argv)


def main(argv: list[str] | None = None) -> int:
    args = parse_args(argv)
    try:
        marker = generate_marker(args.id, args.size, args.border_bits)
        args.output.parent.mkdir(parents=True, exist_ok=True)
        if not cv.imwrite(str(args.output), marker):
            raise OSError(f"OpenCV could not write: {args.output}")
    except (OSError, ValueError, cv.error) as error:
        print(f"Error: {error}", file=sys.stderr)
        return 2
    print(f"Wrote marker {args.id} ({args.size}x{args.size}) to {args.output}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: AugmentedRealityWithArucoMarkers/generate_aruco_markers.cpp
#include <opencv2/core/utility.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/objdetect/aruco_dictionary.hpp>

#include <filesystem>
#include <iostream>
#include <string>

namespace fs = std::filesystem;

const char* kKeys =
    "{help h usage ? |      | Show this message }"
    "{id             |33    | Marker ID from 0 to 249 }"
    "{size           |200   | Square image size in pixels }"
    "{border-bits    |1     | Marker border width in bits }"
    "{output o       |marker33.png| Output PNG path }";

int main(int argc, char** argv) {
    cv::CommandLineParser parser(argc, argv, kKeys);
    parser.about("Generate a DICT_6X6_250 ArUco marker");
    if (parser.has("help")) {
        parser.printMessage();
        return 0;
    }
    if (!parser.check()) {
        parser.printErrors();
        return 2;
    }

    const int marker_id = parser.get<int>("id");
    const int size = parser.get<int>("size");
    const int border_bits = parser.get<int>("border-bits");
    const fs::path output_path = parser.get<std::string>("output");
    if (marker_id < 0 || marker_id >= 250) {
        std::cerr << "Error: --id must be between 0 and 249.\n";
        return 2;
    }
    if (size < 32) {
        std::cerr << "Error: --size must be at least 32 pixels.\n";
        return 2;
    }
    if (border_bits < 1) {
        std::cerr << "Error: --border-bits must be at least 1.\n";
        return 2;
    }

    try {
        const cv::aruco::Dictionary dictionary =
            cv::aruco::getPredefinedDictionary(cv::aruco::DICT_6X6_250);
        cv::Mat marker_image;
        dictionary.generateImageMarker(
            marker_id, size, marker_image, border_bits
        );
        if (output_path.has_parent_path()) {
            fs::create_directories(output_path.parent_path());
        }
        if (!cv::imwrite(output_path.string(), marker_image)) {
            std::cerr << "Error: OpenCV could not write: "
                      << output_path << '\n';
            return 3;
        }
    } catch (const cv::Exception& error) {
        std::cerr << "OpenCV error: " << error.what() << '\n';
        return 3;
    } catch (const fs::filesystem_error& error) {
        std::cerr << "Filesystem error: " << error.what() << '\n';
        return 3;
    }

    std::cout << "Wrote marker " << marker_id << " (" << size << 'x'
              << size << ") to " << output_path << '\n';
    return 0;
}
